godot-core/src/task/futures.rs

```diff
@@ -308,8 +334,188 @@ impl<C: WithBaseField, R: ParamTuple + Sync + Send> IntoFuture for &TypedSignal<
     }
 }
 
+/// Convert a value into a type that is [`Send`] at compile-time while the value might not be.
+///
+/// This allows to turn any implementor into a type that is `Send`, but requires to also implement [`AssertSafeSend`] as well.
+/// The later trait will verify if a value can actually be sent between threads at runtime.
+#[expect(private_bounds)]
+pub trait IntoMaybeSend: Sealed {
+    type Target: Send + AssertSafeSend<Target = Self>;
+
+    fn into_maybe_send(self) -> Self::Target;
+}
+
+/// Assert that it was safe to send the value to the current thread.
+///
+/// # Safety
+/// The implementor has to guarantee that `assert_safe_send` panics if the value has been sent between threads while being `!Send`.
+#[expect(private_bounds)]
+pub unsafe trait AssertSafeSend: Send + Sealed {
+    type Target;
+
+    fn assert_safe_send(self) -> Self::Target;
```

Comment:
Should this maybe be
```rs
fn try_into_send(self) -> Option<Self::Target>;
```
returning `None` if it's forbidden? That makes the fallibility part of the type system, and allows to unify panic messages at the calling side.

Reply:
Then, the trait itself could be named `TryIntoSend`.

Reply:
sounds good.

Reply:
"assert" has the meaning of claim/suggest, which was in line with the panic, but with `Option` it's probably more of a "check". Maybe we should also avoid "[try_]into" as it's not really a conversion but rather a safety check.

`DynamicSend` + `extract_if_safe` or so?